# Actividad R_3 | Concatenación y merge con pandas

## EJERCICIO 1 | Ventas por trimestre

Una empresa tiene las ventas de tres trimestres en DataFrames separados. Hay que unirlos con `pd.concat` y sacar las ventas totales de cada producto.

In [3]:
import pandas as pd

In [4]:
# Creamos un DataFrame por cada trimestre
ventas_q1 = pd.DataFrame({
    "order_id": [101, 102, 103],
    "product_id": [1, 2, 3],
    "product_name": ["Monitor", "Audífonos", "Smartwatch"],
    "units_sold": [6, 15, 8],
    "revenue": [27000, 13500, 20000],
    "quarter": ["Q1", "Q1", "Q1"]
})

ventas_q2 = pd.DataFrame({
    "order_id": [201, 202, 203],
    "product_id": [1, 2, 3],
    "product_name": ["Monitor", "Audífonos", "Smartwatch"],
    "units_sold": [4, 20, 11],
    "revenue": [18000, 18000, 27500],
    "quarter": ["Q2", "Q2", "Q2"]
})

ventas_q3 = pd.DataFrame({
    "order_id": [301, 302, 303],
    "product_id": [1, 2, 3],
    "product_name": ["Monitor", "Audífonos", "Smartwatch"],
    "units_sold": [9, 12, 5],
    "revenue": [40500, 10800, 12500],
    "quarter": ["Q3", "Q3", "Q3"]
})

display(ventas_q1)
display(ventas_q2)
display(ventas_q3)

,order_id,product_id,product_name,units_sold,revenue,quarter
0,101,1,Monitor,6,27000,Q1
1,102,2,Audífonos,15,13500,Q1
2,103,3,Smartwatch,8,20000,Q1


,order_id,product_id,product_name,units_sold,revenue,quarter
0,201,1,Monitor,4,18000,Q2
1,202,2,Audífonos,20,18000,Q2
2,203,3,Smartwatch,11,27500,Q2


,order_id,product_id,product_name,units_sold,revenue,quarter
0,301,1,Monitor,9,40500,Q3
1,302,2,Audífonos,12,10800,Q3
2,303,3,Smartwatch,5,12500,Q3


In [5]:
# Unimos los tres trimestres en un solo DataFrame
# ignore_index=True para que el índice quede corrido del 0 al 8
ventas_trimestres = pd.concat([ventas_q1, ventas_q2, ventas_q3], ignore_index=True)

display(ventas_trimestres)

,order_id,product_id,product_name,units_sold,revenue,quarter
0,101,1,Monitor,6,27000,Q1
1,102,2,Audífonos,15,13500,Q1
2,103,3,Smartwatch,8,20000,Q1
3,201,1,Monitor,4,18000,Q2
4,202,2,Audífonos,20,18000,Q2
5,203,3,Smartwatch,11,27500,Q2
6,301,1,Monitor,9,40500,Q3
7,302,2,Audífonos,12,10800,Q3
8,303,3,Smartwatch,5,12500,Q3


In [6]:
# Filtramos las ventas de cada producto por separado
monitor = ventas_trimestres[ventas_trimestres["product_name"] == "Monitor"]
audifonos = ventas_trimestres[ventas_trimestres["product_name"] == "Audífonos"]
smartwatch = ventas_trimestres[ventas_trimestres["product_name"] == "Smartwatch"]

display(monitor)
display(audifonos)
display(smartwatch)

,order_id,product_id,product_name,units_sold,revenue,quarter
0,101,1,Monitor,6,27000,Q1
3,201,1,Monitor,4,18000,Q2
6,301,1,Monitor,9,40500,Q3


,order_id,product_id,product_name,units_sold,revenue,quarter
1,102,2,Audífonos,15,13500,Q1
4,202,2,Audífonos,20,18000,Q2
7,302,2,Audífonos,12,10800,Q3


,order_id,product_id,product_name,units_sold,revenue,quarter
2,103,3,Smartwatch,8,20000,Q1
5,203,3,Smartwatch,11,27500,Q2
8,303,3,Smartwatch,5,12500,Q3


In [7]:
# Sumamos las unidades vendidas y los ingresos de cada producto
unidades_monitor = monitor["units_sold"].sum()
ingresos_monitor = monitor["revenue"].sum()

unidades_audifonos = audifonos["units_sold"].sum()
ingresos_audifonos = audifonos["revenue"].sum()

unidades_smartwatch = smartwatch["units_sold"].sum()
ingresos_smartwatch = smartwatch["revenue"].sum()

print("Monitor:", unidades_monitor, "unidades -", ingresos_monitor, "de ingresos")
print("Audífonos:", unidades_audifonos, "unidades -", ingresos_audifonos, "de ingresos")
print("Smartwatch:", unidades_smartwatch, "unidades -", ingresos_smartwatch, "de ingresos")

Monitor: 19 unidades - 85500 de ingresos
Audífonos: 47 unidades - 42300 de ingresos
Smartwatch: 24 unidades - 60000 de ingresos


In [8]:
# Armamos el DataFrame final con los totales de cada producto
resultado = pd.DataFrame({
    "product_name": ["Monitor", "Audífonos", "Smartwatch"],
    "total_units": [unidades_monitor, unidades_audifonos, unidades_smartwatch],
    "total_revenue": [ingresos_monitor, ingresos_audifonos, ingresos_smartwatch]
})

# Ordenamos de mayor a menor por unidades vendidas
resultado = resultado.sort_values("total_units", ascending=False)

display(resultado)

,product_name,total_units,total_revenue
1,Audífonos,47,42300
2,Smartwatch,24,60000
0,Monitor,19,85500


## EJERCICIO 2 | Ventas y clientes con merge

Una tienda en línea tiene sus ventas en `ventas.csv` y sus clientes en `clientes.csv`. Hay que unir las dos tablas con `merge()` usando los distintos tipos de join.

In [9]:
# Cargamos los dos archivos csv
ventas = pd.read_csv("ventas.csv")

# clientes.csv marcaba error por un acento guardado en otro formato,
# con encoding_errors="replace" pandas ya lo puede leer
clientes = pd.read_csv("clientes.csv", encoding_errors="replace")

display(ventas)
display(clientes)

,venta_id,cliente_id,producto,cantidad,total_venta
0,101,1,Laptop,1,1200.0
1,102,2,Smartphone,2,1500.0
2,103,3,Tablet,1,600.0
3,104,4,Smartwatch,3,750.0
4,105,5,Laptop,1,1250.0


,cliente_id,nombre,email,ciudad
0,1,Juan Pérez,juan@email.com,CDMX
1,2,Ana Gómez,ana@email.com,Monterrey
2,3,Luis Soto,luis@email.com,Guadalajara
3,4,Carla Ruiz,carla@email.com,M�rida
4,6,Pedro Lara,pedro@email.com,Puebla


In [10]:
# Inner: solo las ventas cuyo cliente sí está registrado
merge_inner = pd.merge(ventas, clientes, on="cliente_id", how="inner")

display(merge_inner)

,venta_id,cliente_id,producto,cantidad,total_venta,nombre,email,ciudad
0,101,1,Laptop,1,1200.0,Juan Pérez,juan@email.com,CDMX
1,102,2,Smartphone,2,1500.0,Ana Gómez,ana@email.com,Monterrey
2,103,3,Tablet,1,600.0,Luis Soto,luis@email.com,Guadalajara
3,104,4,Smartwatch,3,750.0,Carla Ruiz,carla@email.com,M�rida


In [11]:
# Left: todas las ventas, aunque el cliente no esté registrado
merge_left = pd.merge(ventas, clientes, on="cliente_id", how="left")

display(merge_left)

,venta_id,cliente_id,producto,cantidad,total_venta,nombre,email,ciudad
0,101,1,Laptop,1,1200.0,Juan Pérez,juan@email.com,CDMX
1,102,2,Smartphone,2,1500.0,Ana Gómez,ana@email.com,Monterrey
2,103,3,Tablet,1,600.0,Luis Soto,luis@email.com,Guadalajara
3,104,4,Smartwatch,3,750.0,Carla Ruiz,carla@email.com,M�rida
4,105,5,Laptop,1,1250.0,NaN,NaN,NaN


In [12]:
# Right: todos los clientes, aunque no hayan comprado nada
merge_right = pd.merge(ventas, clientes, on="cliente_id", how="right")

display(merge_right)

,venta_id,cliente_id,producto,cantidad,total_venta,nombre,email,ciudad
0,101.0,1,Laptop,1.0,1200.0,Juan Pérez,juan@email.com,CDMX
1,102.0,2,Smartphone,2.0,1500.0,Ana Gómez,ana@email.com,Monterrey
2,103.0,3,Tablet,1.0,600.0,Luis Soto,luis@email.com,Guadalajara
3,104.0,4,Smartwatch,3.0,750.0,Carla Ruiz,carla@email.com,M�rida
4,NaN,6,NaN,NaN,NaN,Pedro Lara,pedro@email.com,Puebla


In [13]:
# Outer: todas las ventas y todos los clientes, coincidan o no
merge_outer = pd.merge(ventas, clientes, on="cliente_id", how="outer")

display(merge_outer)

,venta_id,cliente_id,producto,cantidad,total_venta,nombre,email,ciudad
0,101.0,1,Laptop,1.0,1200.0,Juan Pérez,juan@email.com,CDMX
1,102.0,2,Smartphone,2.0,1500.0,Ana Gómez,ana@email.com,Monterrey
2,103.0,3,Tablet,1.0,600.0,Luis Soto,luis@email.com,Guadalajara
3,104.0,4,Smartwatch,3.0,750.0,Carla Ruiz,carla@email.com,M�rida
4,105.0,5,Laptop,1.0,1250.0,NaN,NaN,NaN
5,NaN,6,NaN,NaN,NaN,Pedro Lara,pedro@email.com,Puebla


### Preguntas

**¿Qué tipo de merge mantuvo solo los clientes con compras?**

El inner. Solo deja las filas donde el `cliente_id` está en las dos tablas, o sea los clientes que sí tienen una venta. Por eso salen 4 filas.

**¿En qué merge aparecen ventas sin clientes registrados?**

En el left (y también en el outer). Como la tabla de la izquierda es `ventas`, se quedan todas las ventas. La venta 105 es del cliente 5, que no está en `clientes.csv`, y por eso sale con NaN en nombre, email y ciudad.

**¿Cuál merge permite ver clientes que no han comprado nada?**

El right (y también el outer). Ahí se quedan todos los clientes. Pedro Lara (cliente 6) sale con NaN en los datos de la venta porque no tiene ninguna compra.

**¿Cómo podrías usar merge para detectar clientes sin compras?**

Haciendo un merge tipo left con `clientes` como tabla de la izquierda y `ventas` a la derecha. Los clientes que no tienen ventas quedan con NaN en `venta_id`, así que solo hay que filtrar esas filas con `isnull()`.

In [14]:
# Clientes que no tienen ninguna venta
clientes_ventas = pd.merge(clientes, ventas, on="cliente_id", how="left")
sin_compras = clientes_ventas[clientes_ventas["venta_id"].isnull()]

display(sin_compras)

,cliente_id,nombre,email,ciudad,venta_id,producto,cantidad,total_venta
4,6,Pedro Lara,pedro@email.com,Puebla,NaN,NaN,NaN,NaN


**¿Cómo podrías identificar ventas sin un cliente registrado?**

Al revés: con el merge left que ya hicimos (`ventas` a la izquierda), las ventas cuyo cliente no existe quedan con NaN en `nombre`. Filtrando esas filas se ven las ventas sin cliente registrado.

In [15]:
# Ventas cuyo cliente no está registrado
sin_cliente = merge_left[merge_left["nombre"].isnull()]

display(sin_cliente)

,venta_id,cliente_id,producto,cantidad,total_venta,nombre,email,ciudad
4,105,5,Laptop,1,1250.0,NaN,NaN,NaN
